pkg/scaffold/internal/filesystem/mock.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filesystem
+
+import (
+	"bytes"
+	"io"
+)
+
+// mockFileSystem implements FileSystem
+type mockFileSystem struct {
+	path            string
+	exists          func(path string) bool
+	existsError     error
+	createDirError  error
+	createFileError error
+	output          *bytes.Buffer
+	writeFileError  error
+	closeFileError  error
+}
+
+// NewMock returns a new FileSystem
+func NewMock(options ...MockOptions) FileSystem {
+	// Default values
+	fs := mockFileSystem{
+		exists: func(_ string) bool { return false },
+		output: new(bytes.Buffer),
+	}
+
+	// Apply options
+	for _, option := range options {
+		option(&fs)
+	}
+
+	return fs
+}
+
+// MockOptions configure FileSystem
+type MockOptions func(system *mockFileSystem)
+
+// MockPath ensures that the file created with this scaffold is at path
+func MockPath(path string) MockOptions {
+	return func(fs *mockFileSystem) {
+		fs.path = path
+	}
+}
+
+// MockExists makes FileSystem.Exists use the provided function to check if the
+// file exists
+func MockExists(exists func(path string) bool) MockOptions {
+	return func(fs *mockFileSystem) {
+		fs.exists = exists
+	}
+}
+
+// MockExistsError makes FileSystem.Exists return err
+func MockExistsError(err error) MockOptions {
+	return func(fs *mockFileSystem) {
```

Comment:
Why we do not use the funcs in `pkg/scaffold/internal/filesystem/errors.go` instead of re-writen them to check the errors?

Reply:
The functions in `pkg/scaffold/internal/filesystem/errors.go` check if an error is of a certain type. This functions make the `mockFileSystem` return an error at that step.

When you call `FileSystem.Exists` an error is returned (it may be `nil`).
- `fileSysten` implements the `FileSystem` interface. If an error happens, for example you are not allowed to read in that directory so you cna't check if it exists, `fileSystem.Exists` will return that error.
- `mockFileSystem` implements the `FileSystem` interface too. When you call `mockFileSystem.Exists` the error returned is the one you provided with `MockCreateFileError`.